pkg/cloudprovider/providers/gce/gce.go

```diff
@@ -172,68 +173,90 @@ func (g *GCECloud) GetProjectID() string {
 }
 
 // newGCECloud creates a new instance of GCECloud.
-func newGCECloud(config io.Reader) (*GCECloud, error) {
-	apiEndpoint := ""
-	projectID, zone, err := getProjectAndZone()
-	if err != nil {
-		return nil, err
-	}
-
-	region, err := GetGCERegion(zone)
-	if err != nil {
-		return nil, err
-	}
-
-	networkName, err := getNetworkNameViaMetadata()
-	if err != nil {
-		return nil, err
-	}
-	networkURL := gceNetworkURL("", projectID, networkName)
-	subnetworkURL := ""
-
-	// By default, Kubernetes clusters only run against one zone
-	managedZones := []string{zone}
-
-	tokenSource := google.ComputeTokenSource("")
+func newGCECloud(config io.Reader) (gceCloud *GCECloud, err error) {
+	var apiEndpoint, projectID, zone, region, networkURL, subnetworkURL string
 	var nodeTags []string
```

Comment:
put the read config in a separate function. return a struct with the attributes that we want to read:

`apiEndpoint, projectID, zone, region, networkURL, subnetworkURL`

write a bunch of unit tests just for the configuration file parsing.